pkg/sql/schemachanger/scdecomp/decomp.go

```diff
@@ -678,3 +676,109 @@ func (w *walkCtx) walkForeignKeyConstraint(
 		})
 	}
 }
+
+func (w *walkCtx) walkFunction(fnDesc catalog.FunctionDescriptor) {
+	typeT, err := newTypeT(fnDesc.GetReturnType().Type)
+	if err != nil {
+		panic(err)
+	}
+	fn := &scpb.Function{
+		FunctionID: fnDesc.GetID(),
+		ReturnSet:  fnDesc.GetReturnType().ReturnSet,
+		ReturnType: *typeT,
+		Params:     make([]scpb.Function_Parameter, len(fnDesc.GetParams())),
+	}
+	for i, param := range fnDesc.GetParams() {
+		typeT, err := newTypeT(param.Type)
+		if err != nil {
+			panic(err)
+		}
+		fn.Params[i] = scpb.Function_Parameter{
+			Name:  param.Name,
+			Class: catpb.FunctionParamClass{Class: param.Class},
+			Type:  *typeT,
+		}
+		if param.DefaultExpr != nil {
+			expr, err := w.newExpression(*param.DefaultExpr)
+			if err != nil {
+				panic(err)
+			}
+			w.ev(scpb.Status_PUBLIC, &scpb.FunctionParamDefaultExpression{
+				FunctionID: fnDesc.GetID(),
+				Ordinal:    uint32(i),
+				Expression: *expr,
+			})
+		}
+	}
+
+	w.ev(descriptorStatus(fnDesc), fn)
+	w.ev(scpb.Status_PUBLIC, &scpb.ObjectParent{
+		ObjectID:       fnDesc.GetID(),
+		ParentSchemaID: fnDesc.GetParentSchemaID(),
+	})
+	w.ev(scpb.Status_PUBLIC, &scpb.FunctionName{
+		FunctionID: fnDesc.GetID(),
+		Name:       fnDesc.GetName(),
+	})
+	w.ev(scpb.Status_PUBLIC, &scpb.FunctionVolatility{
+		FunctionID: fnDesc.GetID(),
+		Volatility: catpb.FunctionVolatility{Volatility: fnDesc.GetVolatility()},
+	})
+	w.ev(scpb.Status_PUBLIC, &scpb.FunctionLeakProof{
+		FunctionID: fnDesc.GetID(),
+		LeakProof:  fnDesc.GetLeakProof(),
+	})
+	w.ev(scpb.Status_PUBLIC, &scpb.FunctionNullInputBehavior{
+		FunctionID:        fnDesc.GetID(),
+		NullInputBehavior: catpb.FunctionNullInputBehavior{NullInputBehavior: fnDesc.GetNullInputBehavior()},
+	})
+
+	fnBody := &scpb.FunctionBody{
+		FunctionID:  fnDesc.GetID(),
+		Body:        fnDesc.GetFunctionBody(),
+		Lang:        catpb.FunctionLanguage{Lang: fnDesc.GetLanguage()},
+		UsesTypeIDs: fnDesc.GetDependsOnTypes(),
+		// TODO(chengxiong): add UsesFunctionIDs
+	}
+	dedupeColIDs := func(colIDs []catid.ColumnID) []catid.ColumnID {
+		ret := catalog.MakeTableColSet()
+		for _, id := range colIDs {
+			ret.Add(id)
+		}
+		return ret.Ordered()
+	}
+	for _, toID := range fnDesc.GetDependsOn() {
+		to := w.lookupFn(toID)
+		toDesc, err := catalog.AsTableDescriptor(to)
+		if err != nil {
+			panic(err)
+		}
```

Comment:
yes, functions can depend on types and that's handled by `UsesTypeIDs: fnDesc.GetDependsOnTypes(),` a few line above. There're cases where tables are used as implicit record types, but we consider that as table dependency instead of type dependency.

Reply:
Thanks for teaching!